runtime/opt/taupage/init.d/04-configure-network-interfaces.py

```diff
@@ -0,0 +1,226 @@
+        super(CmdException, self).__init__(message)
+
+    @property
+    def returncode(self):
+        return self._returncode
+
+
+def call_command(call, allowed_error_codes=[0]):
+    proc = subprocess.Popen(call, stderr=subprocess.PIPE)
+    stdout, stderr = proc.communicate()
+    if proc.returncode not in allowed_error_codes:
+        raise CmdException(proc.returncode, stderr.decode('utf-8'))
+    return stdout
+
+
+def ec2_client(region):
+    return boto.ec2.connect_to_region(region)
+
+
+@retry
+def get_all_network_interfaces(ec2, filters):
+    return ec2.get_all_network_interfaces(filters=filters)
+
+
+def find_network_interface(ec2, name):
+    tries = 10
+    network_interfaces = []
+    while not network_interfaces:
+        try:
+            network_interfaces = list(get_all_network_interfaces(ec2, {
+                'tag:Name': name,
+                'status': 'available',
+                'availability-zone': zone()
+            }))
+        except Exception as e:
+            logging.exception(e)
+            sys.exit(2)
+
+        if not network_interfaces:
+            logging.error(
+                'No matching "available" network interfaces volume with name %s found.', name)
+            tries -= 1
+            if tries > 0:
+                logging.error(
+                    'Sleeping for 10 seconds and hope a network interface will become "available"')
+                time.sleep(10)
+            else:
+                sys.exit(2)
+
+    if len(network_interfaces) > 1:
+        logging.warning('More than one network interface with name %s found.', name)
+        network_interfaces.sort(key=lambda v: v.id)
+    return network_interfaces[0].id
+
+
+def wait_for_network_interface_attachment(ec2, device_index):
+    tries = 3
+
+    def get_attachment():
+        return "eth{}".format(device_index) in netifaces.interfaces()
+
+    attachment = get_attachment()
+    while not attachment:
+        attachment = get_attachment()
+        if not attachment:
+            tries -= 1
+            if tries > 0:
+                logging.error(
+                    'Sleeping for 60 seconds waiting for network interface to attach')
+                time.sleep(60)
+            else:
+                sys.exit(2)
+
+
+@retry
+def attach_network_interface(ec2, network_interface_id, device_index):
+    ec2.attach_network_interface(
+        device_index=device_index,
+        network_interface_id=network_interface_id,
+        instance_id=instance_id()
+    )
+
+
+def handle_network_interfaces(region, config):
+    ec2 = ec2_client(region)
+    network_interfaces = config.get("network_interfaces", [])
+    for index, name in enumerate(network_interfaces):
+        try:
+            device_index = index + 1
+            attach_network_interface(ec2, find_network_interface(ec2, name), device_index)
+            wait_for_network_interface_attachment(ec2, device_index)
```

Comment:
If there are > 1 ec2 instances starting up at the same time (normal behaviour for most ASGs we use) this code runs simultaneously and thus two instances might try to attach the same network interface that they found simultaneously.
One of them will thus fail calling `attach_network_interface`, resulting in the taupage instance not finishing its init process properly: https://github.com/zalando-stups/taupage/blob/master/runtime/opt/taupage/init.sh#L27

I am not sure if this i what you want.

In case `attach_network_interface` fails or `wait_for_network_interface_attachment` fails it might make sense to re-execute: `find_network_interface` to get an available network interface that is not already bound by another ec2 instance, `attach_network_interface` and `wait_for_network_interface_attachment`. (maybe only 3 times)

Reply:
> If there are > 1 ec2 instances starting up at the same time (normal behaviour for most ASGs we use)

I was working under the assumption that it would only be a single instance in the ASG.

Much of the logic here is copied from the mount volume by tag code.

Reply:
If this is fine with other reviewers, then this assumption should be made transparent in the script and docs.

I have to admit that if we could make this script safe for having multiple instances started at the same time going for a group of network_interfaces, with a rather simple change like i suggested above, we should go for it.